flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/AbstractWebSocketHandler.java

```diff
@@ -0,0 +1,301 @@
+		String before = ctx.pipeline().names().get(ctx.pipeline().names().indexOf(ctx.name()) - 1);
+		ChannelHandlerContext beforeCtx = ctx.pipeline().context(before);
+
+		// inject the websocket protocol handler into this channel, to be active
+		// until the channel is closed.  note that the handshake may or may not complete synchronously.
+		ctx.pipeline().addBefore(ctx.name(), WebSocketServerProtocolHandler.class.getName(),
+			new WebSocketServerProtocolHandler(msg.path(), specification.getSubprotocol()));
+
+		// inject the message codec
+		ctx.pipeline().addBefore(ctx.name(), messageCodec.getClass().getName(), messageCodec);
+
+		log.debug("Upgraded channel with WS protocol handler and message codec.");
+
+		// forward the message to the installed protocol handler to initiate handshaking
+		HttpRequest request = msg.request();
+		ReferenceCountUtil.retain(request);
+		beforeCtx.fireChannelRead(request);
+	}
+
+	@Override
+	public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) throws Exception {
+		log.error("WebSocket channel error; closing the channel.", cause);
+		ctx.close();
+	}
+
+	/**
+	 * Handles a client handshake request to open a WebSocket resource.  Returns a {@link CompletableFuture} to complete handshaking.
+	 *
+	 * <p>Implementations may decide whether to throw {@link RestHandlerException}s or fail the returned
+	 * {@link CompletableFuture} with a {@link RestHandlerException}.
+	 *
+	 * <p>Failing the future with another exception type or throwing unchecked exceptions is regarded as an
+	 * implementation error as it does not allow us to provide a meaningful HTTP status code. In this case a
+	 * {@link HttpResponseStatus#INTERNAL_SERVER_ERROR} will be returned.
+	 *
+	 * @param parameters the REST parameters
+
+	 * @return future indicating completion of handshake pre-processing.
+	 * @throws RestHandlerException to produce a pre-formatted HTTP error response.
+	 */
+	protected CompletableFuture<Void> handshakeInitiated(ChannelHandlerContext ctx, M parameters) throws Exception {
+		return CompletableFuture.completedFuture(null);
+	}
+
+	/**
+	 * Invoked when the current channel has completed the handshaking to establish a WebSocket connection.
+	 *
+	 * @param ctx the channel handler context
+	 * @param parameters the REST parameters
+	 * @throws Exception if processing failed.
+	 */
+	protected void handshakeCompleted(ChannelHandlerContext ctx, M parameters) throws Exception {
+	}
+
+	/**
+	 * Invoked when the current channel has received a WebSocket message.
+	 *
+	 * <p>Be sure to release the message object (default behavior).
+	 *
+	 * @param ctx the channel handler context
+	 * @param parameters the REST parameters
+	 * @param msg the message received
+	 * @throws Exception if the message could not be processed.
+	 */
+	protected void messageReceived(ChannelHandlerContext ctx, M parameters, O msg) throws Exception {
+		ReferenceCountUtil.release(msg);
```

Comment:
I think I will change the behavior to be more consistent with `SimpleChannelInboundHandler` and auto-release the message rather than forcing the subclass to release it.  Note that `channelRead0` is being renamed to `messageReceived` in Netty 5.